src/D2L.CodeStyle.Analyzers/Async/Generator/SyntaxTransformer.cs

```diff
@@ -0,0 +1,120 @@
+	protected readonly SemanticModel m_model;
+	private readonly ImmutableArray<Diagnostic>.Builder m_diagnostics = ImmutableArray.CreateBuilder<Diagnostic>();
+
+	/// <summary>
+	/// A name for the generator that will be shown to the user if we crash
+	/// </summary>
+	protected abstract string Name { get; }
+
+	public SyntaxTransformer( SemanticModel model ) {
+		m_model = model;
+	}
+
+	/// <summary>
+	/// Suggested return type for public transformation methods.
+	/// </summary>
+	public sealed record TransformResult<T>(
+		T? Value,
+		ImmutableArray<Diagnostic> Diagnostics
+	) {
+		public bool Success => Diagnostics.Length == 0;
+
+		public TransformResult( ImmutableArray<Diagnostic> diagnostics ) : this(
+			Value: default,
+			Diagnostics: diagnostics
+		) {}
+
+		public TransformResult( T value ) : this(
+			Value: value,
+			Diagnostics: ImmutableArray<Diagnostic>.Empty
+		) { }
+	}
+
+	protected TransformResult<T> GetResult<T>( T value ) {
+		var result = m_diagnostics.ToImmutable();
+		m_diagnostics.Clear();
+
+		if( result.Length == 0 ) {
+			return new( value );
+		}
+
+		return new( result );
+	}
+
+	/// <summary>
+	/// Report a diagnostic and continue.
+	/// </summary>
+	protected void ReportDiagnostic(
+		DiagnosticDescriptor descriptor,
+		Location location,
+		params object[] messageArgs
+	) {
+		var diag = Diagnostic.Create( descriptor, location, messageArgs );
+
+		m_diagnostics.Add( diag );
+	}
+
+	/// <summary>
+	/// Reports an error in our generator to the user.
+	/// </summary>
+	/// <param name="location"></param>
+	/// <param name="what"></param>
+	protected void GeneratorError( Location location, string what ) =>
+		ReportDiagnostic( Diagnostics.GenericGeneratorError, location, Name, what );
+
+	/// <summary>
+	/// Transform every element of a SyntaxList and filter out nulls.
+	/// </summary>
+	protected SyntaxList<U> TransformAll<T, U>(
+		SyntaxList<T> input,
+		Func<T, U?> transformer
+	) where T : SyntaxNode
+	  where U : SyntaxNode
+	=> SyntaxFactory.List(
+		(IEnumerable<U>)input // Cast away the ? because of the Where filter
+			.Select( transformer )
+			.Where( NotNull )
+	);
+
+	/// <summary>
+	/// Transform every element of a SeparatedSyntaxList and filter out nulls.
+	/// </summary>
+	protected SeparatedSyntaxList<U> TransformAll<T, U>(
+		SeparatedSyntaxList<T> input,
+		Func<T, U?> transformer
+	) where T : SyntaxNode
+	  where U : SyntaxNode
+	=> SyntaxFactory.SeparatedList(
+		(IEnumerable<U>)input // Cast away the ? because of the Where filter
+			.Select( transformer )
+			.Where( NotNull )
+	);
+
+	/// <summary>
+	/// Transform every element of a SyntaxTokenList and filter out nulls.
+	/// </summary>
+	protected SyntaxTokenList TransformAll(
+		SyntaxTokenList input,
+		Func<SyntaxToken, SyntaxToken?> transformer
+	) => SyntaxFactory.TokenList(
+		input.Select( transformer )
+			.Where( NotNull )
+			.Select( static t => t!.Value )
+	);
+
+	private static bool NotNull<T>( T? t ) => t != null;

```

Comment:
Take it or leave it. Could theoretically be good for us to use the pattern matching for null checks more generally (no operator overloading), but they're not a tool I'm used to reaching for yet.

```suggestion
	private static bool NotNull<T>( T? t ) => t is not null;
```

Reply:
I don't think I have a preference, this helpers is gone now tho.